### PR TITLE
synchronizing API/PROPERTIES and the actual code

### PR DESCRIPTION
**Pull request by ghost**
This pull-request updates the file `API/PROPERTIES` and makes sure that the actual code is consistent with those statements.

### CODE REVIEW

File: proofs/proof_type.mli [hunk 1]
```diff
@@ -26,3 +26,8 @@ type rule = prim_rule
 type goal = Goal.goal
 
 type tactic = goal sigma -> goal list sigma
+(** @deprecated: Please use "Tacmach.tactic" type instead.
+
+    We keep this definition only for "technical reasons".
+    If one tries to get rid of it, Ocaml compiler complains
+    about "circular dependency" between "Refiner" and "Tacmach" modules. *)
```
Comment (matejkosik) [inline proofs/proof_type.mli]: @ppedrot Is this the correct thing to say?

Previously, I tried to expose this type in the API and you subsequently corrected that mistake. A comment of this sort would help me to remember that this is a bad idea.

**Review by ejgallego — DISMISSED** [review]
I am blocking any further API update until a plan is put in place.

File: proofs/proof_type.mli (hunk 1 above)
Comment (ppedrot) [inline proofs/proof_type.mli]: Technically, the `tactic` type should eventually disappear, so it's not even recommended to switch to the `Tacmach` alias. Furthermore, the `Proof_type` module is probably one of the first modules that are going to disappear when we cut the old engine out of Coq.

Comment (ejgallego) [inline proofs/proof_type.mli]: Note that there is nothing special about this module, however the API itself is declared in the wrong order so of course you have spurious "circular dependencies issues". See #808

**Comment by herbelin** [comment]
@ejgallego: In more constructive terms, what would you concretely expect as "a plan is put in place"?

**Comment by ejgallego** [comment]
@herbelin sorry if I was not being more explicit as we had this discussion at Le Croisic so I was omitting repeating many of the points. By a plan I mean a "list of points" we all @maximedenes @ppedrot @gares @letouzey @matejkosik etc... agree so we can proceed to implement it in an ordered manner instead of doing random fix ups.

I can divide API work in two categories:
- **design questions**: the two main points are the ml vs mli debate and the existence of Grammar_API. IMHO these are blockers to do further work.
- **bugs**: among those are incorrect ordering of modules (#808) , incorrect linking flags / use of -open, and some technical errors in data type declarations. I tried to fix a few in in #808 too.

So yeah, I think consensus on the design questions should be achieved first, once it is in place and the more egregious ordering problems are solved we can proceed to work in parallel.

**Comment by herbelin** [comment]
@ejgallego: Would it be possible to summarize what is consensual at the current time and what is not, including the ml vs mli debate and the grammar_API debate?

**Comment by ejgallego** [comment]
@herbelin AFAICT we have not reached any consensus.

**Comment by ejgallego** [comment]
On the `-open API` vs `open API` I seem to recall that we all but @matejkosik agreed on the change, but I guess it should be confirmed. @matejkosik stated that he didn't understand the reason behind the change.

The rest are correctness issues so IMHO the fixes are clear.

**Comment by ejgallego** [comment]
@herbelin et al, I am not sure what would be the best forum to discuss the design issues; maybe github is not "real-time" enough, we could try gitter or even an `appear.in` if people want.

My view on the two design questions is:
- We should get rid of `.mli`-only module files, and indeed of `intf/`. I tried to ask for advice upstream [https://discuss.ocaml.org/t/future-of-mli-only-module-setups/445] but it seems that aliases for mli-only modules are not really supported by OCaml. @letouzey stated that he preferred we stay `mli`-only, but then, I don't see how the API would be possible. Another workaround would be to use some module magic to generate an implementation in the alias module (in the style of #809 ) but that requires Ocaml >= 4.03.

- Regarding the `Grammar_API`, my position is that such module is not useful (many functions there are really used by plugins) and creates linking problems as there is a mutual dependency with `API.mli`. IMHO such refinement is a case of premature optimization.

**Comment by herbelin** [comment]
@ejgallego:
- Both of `ml` and `mli` seem to work in practice, independently of the advantage of using `ml` for `API`, right? Are there race conditions with using `ml`, or other drawbacks in using `ml`?
- Is the question to have `Grammar_API` or integrated to `API`? What would it bring to keep the former separated?

**Comment by ejgallego** [comment]
> Both of ml and mli seem to work in practice, independently of the advantage of using ml for API, right? Are there race conditions with using ml, or other drawbacks in using ml?

What do you mean by "both". A mli-only setup doesn't allow aliasing modules so it is indeed a problem. A ml-only setup is, in my opinion, undesirable but could work, however it seems there are problem with Coq's make system. I ignore if these problems are particular to Coq or highlight a deeper issue.

> Is the question to have Grammar_API or integrated to API? What would it bring to keep the former separated? [Edited: apparently #854 addresses this latter issue.]

There are two issues with `Grammar_API`:
- it would be OK to have an API that is accessible only to `.ml4` files, however, this is not the case, as the `Grammar_API` is accessed and used by `ml` files; also, some components such as the parser have to live there due to the next restriction.
- As I pointed out the current set interacts badly with the linking structure in the sense of mutual dependencies. This could be sorted out but would likely require many modules.
- due to 1, it requires modules to do `open Grammar_API`; sorry, this is no-go for me. Whether the API is in use should be a configure-time decision and not hardcoded into the source file.

Overall, I see no benefit and some serious problems, but I'd like to hear what others think.

File: API/PROPERTIES [hunk 2]
```diff
@@ -33,6 +33,11 @@
     Until we find the answer to those questions (for each and every alias)
     we will continue to regard aliases as noise and thus remove/deprecate them.
 
+    CONSEQUENCE
+
+    "Constr" module does not appear in the API because whatever it provides
+    is also provided by "Term" (which is already present in the API).
+
```
Comment (matejkosik) [inline API/PROPERTIES]: This is in conflict with PR #808 which (among many good ideas) also unfortunately repopulates API with many aliases.

Comment (ejgallego) [inline API/PROPERTIES]: Indeed I think policy has to be updated; API should provide compatibility aliases and all the aliases that are present in the base coq interfaces. If an alias is to be removed, it should be first done in the base Coq, then allow some time for the plugin writers to adapt to it.

**Pull request closed without merging**